deploy-queue/src/handler/mod.rs

```diff
@@ -35,11 +35,52 @@ pub async fn enqueue_deployment(client: &Pool<Postgres>, deployment: Deployment)
     Ok(deployment_id)
 }
 
+/// Cancel deployments with stale heartbeats
+async fn cancel_stale_heartbeat_deployments(client: &Pool<Postgres>) -> Result<()> {
+    let heartbeat_timeout_interval = HEARTBEAT_TIMEOUT.to_pg_interval()?;
+
+    let stale_deployments = sqlx::query!(
+        r#"
+        SELECT id, component, version, heartbeat_timestamp
+        FROM deployments
+        WHERE heartbeat_timestamp IS NOT NULL
+          AND finish_timestamp IS NULL
+          AND cancellation_timestamp IS NULL
+          AND heartbeat_timestamp < NOW() - $1::interval
+        "#,
+        heartbeat_timeout_interval
+    )
+    .fetch_all(client)
+    .await?;
+
+    for deployment in stale_deployments {
+        log::warn!(
+            "Cancelling deployment {} ({}, version={}) due to stale heartbeat (last seen: {:?})",
```

Comment:
this is where I'd log the delay as well, something like `last seen <interval> ago: <timestamp>`, utilizing humantime to make the interval human readable, similar to ETAs.